src/combinatorics/simple_graph/partition.lean

```diff
@@ -0,0 +1,151 @@
+/-
+Copyright (c) 2021 Arthur Paulino. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Kyle Miller
+-/
+
+import combinatorics.simple_graph.coloring
+
+/-!
+# Graph partitions
+
+This module provides an interface for dealing with partitions on simple graphs. A partition of
+a graph `G`, with vertices `V`, is a set `P` of disjoint nonempty subsets of `V` such that:
+
+* The union of the subsets in `P` is `V`.
+
+* * Each element of `P` is an independent set. (Each subset contains no pair of adjacent vertices.)
+
+Graph partitions are graph colorings that do not name their colors.  They are adjoint in the
+following sense. Given a graph coloring, there is an associated partition from the set of color
+classes, and given a partition, there is an associated graph coloring from using the partition's
+subsets as colors.  Going from graph colorings to partitions and back makes a coloring "canonical":
+all colors are given a canonical name and unused colors are removed.  Going from partitions to
+graph colorings and back is the identity.
+
+## Main definitions
+
+* `partition` is a structure to represent a partition of a simple graph
+
+* `partition.to_coloring` and `partition.to_coloring'` are functions that create colorings from
+  partitions
+
+* `coloring.to_partition` is a function that creates a partition from a coloring
+
+## Todo
+
+* Define k-partite graphs
+
+* Prove that k-partite graphs are k-colorable and vice-versa
+
+-/
+
+universes u v
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/--
+A `partition` of a simple graph `G` is a structure constituted by
+* `parts`: a set of subsets of the vertices `V` of `G`
+* `is_partition`: a proof that `parts` is a proper partition of `V`
+* `independent`: a proof that each element of `parts` doesn't have a pair of adjacent vertices
+-/
+structure partition :=
+(parts : set (set V))
+(is_partition : setoid.is_partition parts)
+(independent : ∀ (s ∈ parts), is_antichain G.adj s)
+
+/-- Whether a graph can be partitioned in at most `n` parts. -/
+def partitionable (n : ℕ) : Prop :=
+∃ (P : G.partition), nonempty (P.parts ↪ fin n)
+
+namespace partition
+variables {G} (P : G.partition)
+
+/-- Whether a partition `P` has at most `n` parts. -/
+def parts_card_le (h : P.parts.finite) (n : ℕ) : Prop := h.to_finset.card ≤ n
```

Comment:
This might be more convenient, so you don't have to supply the finiteness proof when you use it as a hypothesis:
```suggestion
def parts_card_le (n : ℕ) : Prop := ∃ (h : P.parts.finite), h.to_finset.card ≤ n
```